File: javascript/test/errors.test.ts
import assert from 'assert'
import fs from 'fs'
import yaml from 'js-yaml'

import parse from '../src/index.js'
import { testDataDir } from './testDataDir.js'

type ErrorTest = {
  expression: string
  error: string
}

const tests = yaml.load(fs.readFileSync(`${testDataDir}/errors.yml`, 'utf-8')) as ErrorTest[]

describe('Errors', () => {
  for (const test of tests) {
    it(`fails to parse "${test.expression}" with "${test.error}"`, () => {
      assert.throws(() => parse(test.expression), { message: test.error })
    })
  }
})

File: go/errors_test.go
package tagexpressions

import (
	"fmt"
	"gopkg.in/yaml.v3"
	"io/ioutil"
	"testing"

	"github.com/stretchr/testify/require"
)

type ErrorTest struct {
	Expression string `yaml:"expression"`
	Error      string `yaml:"error"`
}

func TestErrors(t *testing.T) {
	contents, err := ioutil.ReadFile("../testdata/errors.yml")
	require.NoError(t, err)
	var tests []ErrorTest
	err = yaml.Unmarshal(contents, &tests)
	require.NoError(t, err)

	for _, test := range tests {
		name := fmt.Sprintf("fails to parse \"%s\" with \"%s\"", test.Expression, test.Error)
		t.Run(name, func(t *testing.T) {
			_, err := Parse(test.Expression)
			require.Error(t, err)
			require.Equal(t, test.Error, err.Error())
		})
	}
}
